caffe2/onnx/backend.cc

```diff
@@ -674,25 +674,40 @@ Caffe2Ops Caffe2Backend::CreateGemm(
 
   auto trans_a = onnx_node->attributes.get<int64_t>("transA", 0L);
   auto trans_b = onnx_node->attributes.get<int64_t>("transB", 0L);
-  auto broadcast = onnx_node->attributes.get<int64_t>("broadcast", 0L);
+  // Support broadcast by default when opset_version > 6.
+  auto broadcast =
+    onnx_node->attributes.get<int64_t>("broadcast",
+                                       (ctx.opset_version() > 6) ? 1L : 0L);
 
   bool use_fc = false;
-  if ((!trans_a) && trans_b) {
-    if (broadcast) {
-      use_fc = true;
-    } else {
-      const auto input_c_vi_iter = ctx.value_infos().find(node.input(2));
-      if (input_c_vi_iter != ctx.value_infos().end() &&
-          input_c_vi_iter->second.type().tensor_type().shape().dim_size() ==
-              1) {
-        use_fc = true;
+  if (!trans_a && broadcast) {
+    // If we have broadcast and the input_c's dim matching the FC requirement,
+    // we could convert this onnx op to caffe2 FC op.
+    const auto input_c_vi_iter = ctx.value_infos().find(node.input(2));
+    const auto input_b_vi_iter = ctx.value_infos().find(node.input(1));
+    if (input_c_vi_iter != ctx.value_infos().end() &&
+        input_c_vi_iter->second.type().tensor_type().has_shape() &&
+        input_b_vi_iter != ctx.value_infos().end() &&
```

Comment:
This new code will only try to use FC when input_b's shape information is available, which is not desired.

Reply:
I think in real world models, input b and c are usually weights, so we should be able to optimize them.

Reply:
no gemm is pretty common, so you don't know whether a, b and c are inputs or weights.

Reply:
@JerryShih let's be more aggressive, two things to do:
1) report error if the invalid condition appear,
2) if B's shape information is not available, always assume shapes match.

Reply:
How about these two case if c is a scalar?
a)

        A = np.random.randn(1, 3).astype(np.float32)
        B = np.random.randn(3, 1).astype(np.float32)
        C = np.random.randn(1).astype(np.float32)
        node_def = make_node(
            'Gemm',
            ['A', 'B', 'C'],
            ["Y"])
        output = c2.run_node(node_def, [A, B, C])
        np.testing.assert_almost_equal(output["Y"], np.dot(A, B) + C)

b)

        A = np.random.randn(3, 2).astype(np.float32)
        B = np.random.randn(2, 4).astype(np.float32)
        C = np.random.randn(1).astype(np.float32)
        node_def = make_node(
            'Gemm',
            ['A', 'B', 'C'],
            ["Y"])
        output = c2.run_node(node_def, [A, B, C])
        np.testing.assert_almost_equal(output["Y"], np.dot(A, B) + C)

The a) and b) could both use "MatMul+add".
The a) can use FC, but b) can't.
If there is no B's shape information, we can't recognize which case we have now.

@bddppq @houseroad

Reply:
If B's shape information is not available, treating scalar as `MatMul+Add` is fine.